activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -9,7 +9,7 @@ module ActiveRecord
   module ConnectionAdapters
     module AbstractPool # :nodoc:
       def get_schema_cache(connection)
-        self.schema_cache ||= SchemaCache.new(connection)
+        self.schema_cache ||= defined?(PostgreSQLAdapter) && connection.kind_of?(PostgreSQLAdapter) ? PostgreSQL::SchemaCache.new(connection) : SchemaCache.new(connection)
```

Comment:
I think instead of making this a check for adapter type, you could add a method to the adapter that initializes the schema cache for that adapter. Then you can make the `PostgreSQLAdapter` override the default method and provide its specific `PostgreSQL::SchemaCache`